In [3]:
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import pandas as pd


In [4]:
penguins = pd.read_csv('datasets\\penguins.csv')

In [5]:
penguins.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   rowid              344 non-null    int64  
 1   species            344 non-null    object 
 2   island             344 non-null    object 
 3   bill_length_mm     342 non-null    float64
 4   bill_depth_mm      342 non-null    float64
 5   flipper_length_mm  342 non-null    float64
 6   body_mass_g        342 non-null    float64
 7   sex                333 non-null    object 
 8   year               344 non-null    int64  
dtypes: float64(4), int64(2), object(3)
memory usage: 24.3+ KB


In [6]:
penguins.describe()

rowid  bill_length_mm  bill_depth_mm  flipper_length_mm  \
count  344.000000      342.000000     342.000000         342.000000   
mean   172.500000       43.921930      17.151170         200.915205   
std     99.448479        5.459584       1.974793          14.061714   
min      1.000000       32.100000      13.100000         172.000000   
25%     86.750000       39.225000      15.600000         190.000000   
50%    172.500000       44.450000      17.300000         197.000000   
75%    258.250000       48.500000      18.700000         213.000000   
max    344.000000       59.600000      21.500000         231.000000   

       body_mass_g         year  
count   342.000000   344.000000  
mean   4201.754386  2008.029070  
std     801.954536     0.818356  
min    2700.000000  2007.000000  
25%    3550.000000  2007.000000  
50%    4050.000000  2008.000000  
75%    4750.000000  2009.000000  
max    6300.000000  2009.000000

In [7]:
penguins.isnull().sum()

rowid                 0
species               0
island                0
bill_length_mm        2
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
year                  0
dtype: int64

In [8]:
penguins['bill_length_mm']=penguins['bill_length_mm'].fillna(penguins['bill_length_mm'].mean())

In [9]:
penguins['bill_depth_mm']=penguins['bill_depth_mm'].fillna(penguins['bill_depth_mm'].mean())

In [10]:
penguins['flipper_length_mm']=penguins['flipper_length_mm'].fillna(penguins['flipper_length_mm'].mean())

In [11]:
penguins['body_mass_g']=penguins['body_mass_g'].fillna(penguins['body_mass_g'].mean())

In [12]:
penguins['sex']=penguins['sex'].fillna(penguins['sex'].mode()[0])

In [13]:
penguins.isnull().sum()

rowid                0
species              0
island               0
bill_length_mm       0
bill_depth_mm        0
flipper_length_mm    0
body_mass_g          0
sex                  0
year                 0
dtype: int64

In [14]:
penguins['bill_depth_mm']=penguins['bill_depth_mm'].fillna(penguins['bill_depth_mm'].mean())

In [15]:
penguins.isnull().sum()

rowid                0
species              0
island               0
bill_length_mm       0
bill_depth_mm        0
flipper_length_mm    0
body_mass_g          0
sex                  0
year                 0
dtype: int64

In [16]:
penguins['species']== 'Adelie'

0       True
1       True
2       True
3       True
4       True
       ...  
339    False
340    False
341    False
342    False
343    False
Name: species, Length: 344, dtype: bool

In [17]:
penguins[penguins['species']== 'Adelie']

rowid species     island  bill_length_mm  bill_depth_mm  \
0        1  Adelie  Torgersen        39.10000       18.70000   
1        2  Adelie  Torgersen        39.50000       17.40000   
2        3  Adelie  Torgersen        40.30000       18.00000   
3        4  Adelie  Torgersen        43.92193       17.15117   
4        5  Adelie  Torgersen        36.70000       19.30000   
..     ...     ...        ...             ...            ...   
147    148  Adelie      Dream        36.60000       18.40000   
148    149  Adelie      Dream        36.00000       17.80000   
149    150  Adelie      Dream        37.80000       18.10000   
150    151  Adelie      Dream        36.00000       17.10000   
151    152  Adelie      Dream        41.50000       18.50000   

     flipper_length_mm  body_mass_g     sex  year  
0           181.000000  3750.000000    male  2007  
1           186.000000  3800.000000  female  2007  
2           195.000000  3250.000000  female  2007  
3           200.915205  4201.754386    male  2007  
4           193.000000  3450.000000  female  2007  
..                 ...          ...     ...   ...  
147         184.000000  3475.000000  female  2009  
148         195.000000  3450.000000  female  2009  
149         193.000000  3750.000000    male  2009  
150         187.000000  3700.000000  female  2009  
151         201.000000  4000.000000    male  2009  

[152 rows x 9 columns]

<h1> Does a specie live exclusively on an island<h1>

In [18]:
penguins['island'] == 'Biscoe'

0      False
1      False
2      False
3      False
4      False
       ...  
339    False
340    False
341    False
342    False
343    False
Name: island, Length: 344, dtype: bool

In [19]:
(penguins['species']== 'Adelie') & (penguins['island'] == 'Biscoe')

0      False
1      False
2      False
3      False
4      False
       ...  
339    False
340    False
341    False
342    False
343    False
Length: 344, dtype: bool

<h1>Biscoe island, adelie specie population</h1>

In [20]:
penguins[(penguins['species']== 'Adelie') & (penguins['island'] == 'Biscoe')].count()

rowid                44
species              44
island               44
bill_length_mm       44
bill_depth_mm        44
flipper_length_mm    44
body_mass_g          44
sex                  44
year                 44
dtype: int64

<h1>Torgersen island, adelie specie population</h1>

In [21]:
penguins[(penguins['species']== 'Adelie') & (penguins['island'] == 'Torgersen')].count()

rowid                52
species              52
island               52
bill_length_mm       52
bill_depth_mm        52
flipper_length_mm    52
body_mass_g          52
sex                  52
year                 52
dtype: int64

<h1>Dream island, adelie specie population</h1>

In [22]:
penguins[(penguins['species']== 'Adelie') & (penguins['island'] == 'Dream')].count()

rowid                56
species              56
island               56
bill_length_mm       56
bill_depth_mm        56
flipper_length_mm    56
body_mass_g          56
sex                  56
year                 56
dtype: int64

<h1>Insight#1</h1>
<h2>Adelie specie is spread among the islands</h2>

In [23]:
penguins['species'].value_counts()

species
Adelie       152
Gentoo       124
Chinstrap     68
Name: count, dtype: int64

<h1>For Gentoo specie<h1>

<h1>Gentoo that lives on Biscoe</h1>

In [24]:
penguins[(penguins['species']== 'Gentoo') & (penguins['island'] == 'Biscoe')].count()

rowid                124
species              124
island               124
bill_length_mm       124
bill_depth_mm        124
flipper_length_mm    124
body_mass_g          124
sex                  124
year                 124
dtype: int64

<h1>Gentoo that lives on torgersen</h1>

In [25]:
penguins[(penguins['species']== 'Gentoo') & (penguins['island'] == 'Torgersen')].count()

rowid                0
species              0
island               0
bill_length_mm       0
bill_depth_mm        0
flipper_length_mm    0
body_mass_g          0
sex                  0
year                 0
dtype: int64

<h1>Gentoo that lives on Dream</h1>

In [26]:
penguins[(penguins['species']== 'Gentoo') & (penguins['island'] == 'Dream')].count()

rowid                0
species              0
island               0
bill_length_mm       0
bill_depth_mm        0
flipper_length_mm    0
body_mass_g          0
sex                  0
year                 0
dtype: int64

<h1>Insight#2<h1>
<h2>Gentoo lives on biscoe only<h2>

<h1>For chinstrap specie<h1>

<h1>chinstrap that lives on Biscoe</h1> 

In [27]:
penguins[(penguins['species']== 'Chinstrap') & (penguins['island'] == 'Biscoe')].count()

rowid                0
species              0
island               0
bill_length_mm       0
bill_depth_mm        0
flipper_length_mm    0
body_mass_g          0
sex                  0
year                 0
dtype: int64

<h1>chinstrap that lives on Torgersen</h1> 

In [28]:
penguins[(penguins['species']== 'Chinstrap') & (penguins['island'] == 'Torgersen')].count()

rowid                0
species              0
island               0
bill_length_mm       0
bill_depth_mm        0
flipper_length_mm    0
body_mass_g          0
sex                  0
year                 0
dtype: int64

<h1>chinstrap that lives on Dream</h1> 

In [29]:
penguins[(penguins['species']== 'Chinstrap') & (penguins['island'] == 'Dream')].count()

rowid                68
species              68
island               68
bill_length_mm       68
bill_depth_mm        68
flipper_length_mm    68
body_mass_g          68
sex                  68
year                 68
dtype: int64

<h1>Insight#3<h1>
<h2>Chinstrap lives only on Dream island</h2>

<h1>Does the island affect the weight of the specie?</h1>

In [30]:
chinstrap_dream=penguins[(penguins['species']== 'Chinstrap') & (penguins['island'] == 'Dream')]

In [31]:
chinstrap_torg=penguins[(penguins['species']== 'Chinstrap') & (penguins['island'] == 'Torgersen')]

In [32]:
chinstrap_bisc=penguins[(penguins['species']== 'Chinstrap') & (penguins['island'] == 'Biscoe')]

In [33]:
chinstrap_dream['body_mass_g']

276    3500.0
277    3900.0
278    3650.0
279    3525.0
280    3725.0
        ...  
339    4000.0
340    3400.0
341    3775.0
342    4100.0
343    3775.0
Name: body_mass_g, Length: 68, dtype: float64

In [34]:
np.round(chinstrap_dream['body_mass_g'].mean(),1)/1000

np.float64(3.7331)

<h1>Insight#4</h1>
<h2>Chinstrap that lives on dream weighs about 3.73 kl</h2>

In [35]:
np.round(chinstrap_bisc['body_mass_g'].mean())/1000

np.float64(nan)

In [36]:
np.round(chinstrap_torg['body_mass_g'].mean())/1000

np.float64(nan)

In [37]:
adelie_bisc=penguins[(penguins['species']== 'Adelie') & (penguins['island'] == 'Biscoe')]

In [38]:
Adelie_torg=penguins[(penguins['species']== 'Adelie') & (penguins['island'] == 'Torgersen')]

In [39]:
Adelie_dream=penguins[(penguins['species']== 'Adelie') & (penguins['island'] == 'Dream')]

In [40]:
np.round(adelie_bisc['body_mass_g'].mean())/1000

np.float64(3.71)

In [41]:
np.round(Adelie_torg['body_mass_g'].mean())/1000

np.float64(3.716)

In [42]:
np.round(Adelie_dream['body_mass_g'].mean())/1000

np.float64(3.688)

<h1>Insight#5</h1>
<h2>Adelie weights 3.7kl regardless of what island</h1>

In [43]:
Gentoo_bisc=penguins[(penguins['species']== 'Gentoo') & (penguins['island'] == 'Biscoe')]

In [44]:
Gentoo_torg=penguins[(penguins['species']== 'Gentoo') & (penguins['island'] == 'Torgersen')]

In [45]:
Gentoo_dream=penguins[(penguins['species']== 'Gentoo') & (penguins['island'] == 'Dream')]

In [46]:
np.round(Gentoo_bisc['body_mass_g'].mean())/1000

np.float64(5.069)

In [47]:
np.round(Gentoo_torg['body_mass_g'].mean())/1000

np.float64(nan)

In [48]:
np.round(Gentoo_dream['body_mass_g'].mean())/1000

np.float64(nan)

<h1>Insight#6</h1>
<h1>Gentoo that lives on dream weighs about 5.0 kl</h1>

<h1>
    weight per specie</h1>

In [49]:
Gentoo_weight=penguins[(penguins['species']== 'Gentoo')]

In [50]:
adelie_weight=penguins[(penguins['species']== 'Adelie')]

In [51]:
chinstrap_weight=penguins[(penguins['species']== 'Chinstrap')]

In [52]:
np.round(Gentoo_weight['body_mass_g'].mean())/1000

np.float64(5.069)

<h3>The average weight of Gentoo is 5.06</h3>

In [53]:
np.round(adelie_weight['body_mass_g'].mean())/1000

np.float64(3.704)

<h3>The average weight of Adelie is 3.7</h3>

In [54]:
np.round(chinstrap_weight['body_mass_g'].mean())/1000

np.float64(3.733)

<h3>The average weight of Chinstrap is 3.73</h3>

<h1>Insight #7<br><br>Heaviest specie of penguin is Gentoo</h1>

<h1>What year has the most recorded male adelie?</h1> 

In [55]:
penguins_agg = penguins.groupby(['species','island','sex']).count()

In [56]:
penguins_agg

rowid  bill_length_mm  bill_depth_mm  \
species   island    sex                                            
Adelie    Biscoe    female     22              22             22   
                    male       22              22             22   
          Dream     female     27              27             27   
                    male       29              29             29   
          Torgersen female     24              24             24   
                    male       28              28             28   
Chinstrap Dream     female     34              34             34   
                    male       34              34             34   
Gentoo    Biscoe    female     58              58             58   
                    male       66              66             66   

                            flipper_length_mm  body_mass_g  year  
species   island    sex                                           
Adelie    Biscoe    female                 22           22    22  
                    male                   22           22    22  
          Dream     female                 27           27    27  
                    male                   29           29    29  
          Torgersen female                 24           24    24  
                    male                   28           28    28  
Chinstrap Dream     female                 34           34    34  
                    male                   34           34    34  
Gentoo    Biscoe    female                 58           58    58  
                    male                   66           66    66

In [59]:
adelie_male = penguins.groupby(['species','sex','year']).size()

In [60]:
adelie_male

species    sex     year
Adelie     female  2007    22
                   2008    25
                   2009    26
           male    2007    28
                   2008    25
                   2009    26
Chinstrap  female  2007    13
                   2008     9
                   2009    12
           male    2007    13
                   2008     9
                   2009    12
Gentoo     female  2007    16
                   2008    22
                   2009    20
           male    2007    18
                   2008    24
                   2009    24
dtype: int64

In [61]:
adelie_male.unstack()

year              2007  2008  2009
species   sex                     
Adelie    female    22    25    26
          male      28    25    26
Chinstrap female    13     9    12
          male      13     9    12
Gentoo    female    16    22    20
          male      18    24    24

<h1>Insight#8</h1>
<h2>The most adelie male was recorded in 2007</h2>

<h1>Question 9</h1>
<h3>What species weighs more per sex? </h3>

In [62]:
weight_sex = np.round(penguins.groupby(['species','sex'])['body_mass_g'].mean()/1000,1)

In [63]:
weight_sex

species    sex   
Adelie     female    3.4
           male      4.0
Chinstrap  female    3.5
           male      3.9
Gentoo     female    4.7
           male      5.4
Name: body_mass_g, dtype: float64

In [64]:
weight_sex.unstack()

sex        female  male
species                
Adelie        3.4   4.0
Chinstrap     3.5   3.9
Gentoo        4.7   5.4

<h1>Insight #9</h1>
<h3>Gentoo is the heaviest specie per sex</h3>

<h1>Question 10</h1>
<h3>What year is the highest recorded male and female weight? </h3>

In [65]:
weight_sexperyr = np.round(penguins.groupby(['species','sex','year'])['body_mass_g'].mean()/1000,1)

In [66]:
weight_sexperyr.unstack()

year              2007  2008  2009
species   sex                     
Adelie    female   3.4   3.4   3.3
          male     4.0   4.1   4.0
Chinstrap female   3.6   3.5   3.5
          male     3.8   4.1   3.9
Gentoo    female   4.6   4.6   4.8
          male     5.5   5.4   5.4

<h1>Insight #10</h1>
<h3>Gentoo is the heaviest specie with 4.8 for female in 2009, while in male 5.5 is recorded in 2007</h3>

<h1>Question 11</h1>
<h3>Correlation between bill length and bill depth</h3>

<h1>Statistical treatment</h1>
<h3>Correlation using Pearson r</h3>

<h1>P-value</h1>
<h3>A P-value with 0.5 above suggests correlation while below 0.5 shows less or no correlation</h3>

In [67]:
b_length = penguins['bill_length_mm'].corr(penguins['bill_depth_mm'])

In [68]:
b_length

np.float64(-0.23505287035553282)

<h1>Insight #11</h1>
<h3>There is no Correlation between bill length and bill depth</h3>

In [69]:
penguins

rowid    species     island  bill_length_mm  bill_depth_mm  \
0        1     Adelie  Torgersen        39.10000       18.70000   
1        2     Adelie  Torgersen        39.50000       17.40000   
2        3     Adelie  Torgersen        40.30000       18.00000   
3        4     Adelie  Torgersen        43.92193       17.15117   
4        5     Adelie  Torgersen        36.70000       19.30000   
..     ...        ...        ...             ...            ...   
339    340  Chinstrap      Dream        55.80000       19.80000   
340    341  Chinstrap      Dream        43.50000       18.10000   
341    342  Chinstrap      Dream        49.60000       18.20000   
342    343  Chinstrap      Dream        50.80000       19.00000   
343    344  Chinstrap      Dream        50.20000       18.70000   

     flipper_length_mm  body_mass_g     sex  year  
0           181.000000  3750.000000    male  2007  
1           186.000000  3800.000000  female  2007  
2           195.000000  3250.000000  female  2007  
3           200.915205  4201.754386    male  2007  
4           193.000000  3450.000000  female  2007  
..                 ...          ...     ...   ...  
339         207.000000  4000.000000    male  2009  
340         202.000000  3400.000000  female  2009  
341         193.000000  3775.000000    male  2009  
342         210.000000  4100.000000    male  2009  
343         198.000000  3775.000000  female  2009  

[344 rows x 9 columns]

<h1>Question 12</h1>
<h3>Correlation between body mass and flipper length</h3>

<h1>Statistical treatment</h1>
<h3>Correlation using Pearson r</h3>

In [70]:
f_length = penguins['body_mass_g'].corr(penguins['flipper_length_mm'])

In [71]:
f_length

np.float64(0.8712017673060114)

<h1>Insight #12</h1>
<h3>There is positive Correlation between body mass and flipper length</h3>

<h1>Question 13</h1>
<h3>Correlation between bill depth and body mass</h3>

<h1>Statistical treatment</h1>
<h3>Correlation using Pearson r</h3>

In [72]:
mass_depth = penguins['body_mass_g'].corr(penguins['bill_depth_mm'])

In [73]:
mass_depth

np.float64(-0.47191562118606645)

<h1>Insight #13</h1>
<h3>There is no Correlation between body mass and bill depth</h3>

<h1>Question 14</h1>
<h3>Correlation between flipper lenth and bill length</h3>

<h1>Statistical treatment</h1>
<h3>Correlation using Pearson r</h3>

In [74]:
bill_flip = penguins['bill_length_mm'].corr(penguins['flipper_length_mm'])

In [75]:
bill_flip

np.float64(0.6561813407464278)

<h1>Insight #14</h1>
<h3>There is positive Correlation between bill length and flipper length</h3>

In [76]:
pip install scipy


Defaulting to user installation because normal site-packages is not writeableNote: you may need to restart the kernel to use updated packages.




[notice] A new release of pip is available: 24.3.1 -> 25.0.1
[notice] To update, run: C:\Users\LAB4-StudentPC33\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [80]:
from scipy.stats import f_oneway

<h1>Question # 15</h1>
 <h2>ANOVA TEST FOR Male and female based on Body Mass</h2>

In [83]:

f_stats,p_value = f_oneway(penguins[penguins['sex']== 'male']['body_mass_g'],
penguins[penguins['sex']== 'female']['body_mass_g'])

p_value

np.float64(3.0226600487050246e-15)

<h1>Insight # 15</h1>
 <h2>ANOVA test for Male and female based on Body Mass suggests STRONG CORRELATION</h2>

<h1>Question # 16</h1>
 <h2>ANOVA TEST FOR Adelie and Chinstrap based on bill length</h2>

In [86]:

f_stats,p_value = f_oneway(penguins[penguins['species']== 'Adelie']['bill_length_mm'],
penguins[penguins['species']== 'Chinstrap']['bill_length_mm'])

p_value

np.float64(4.724199491448494e-62)

<h1>Insight # 15</h1>
 <h2>ANOVA test for Adelie and Chinstrap based on Bill length suggests STRONG CORRELATION</h2>

In [79]:
penguins

rowid    species     island  bill_length_mm  bill_depth_mm  \
0        1     Adelie  Torgersen        39.10000       18.70000   
1        2     Adelie  Torgersen        39.50000       17.40000   
2        3     Adelie  Torgersen        40.30000       18.00000   
3        4     Adelie  Torgersen        43.92193       17.15117   
4        5     Adelie  Torgersen        36.70000       19.30000   
..     ...        ...        ...             ...            ...   
339    340  Chinstrap      Dream        55.80000       19.80000   
340    341  Chinstrap      Dream        43.50000       18.10000   
341    342  Chinstrap      Dream        49.60000       18.20000   
342    343  Chinstrap      Dream        50.80000       19.00000   
343    344  Chinstrap      Dream        50.20000       18.70000   

     flipper_length_mm  body_mass_g     sex  year  
0           181.000000  3750.000000    male  2007  
1           186.000000  3800.000000  female  2007  
2           195.000000  3250.000000  female  2007  
3           200.915205  4201.754386    male  2007  
4           193.000000  3450.000000  female  2007  
..                 ...          ...     ...   ...  
339         207.000000  4000.000000    male  2009  
340         202.000000  3400.000000  female  2009  
341         193.000000  3775.000000    male  2009  
342         210.000000  4100.000000    male  2009  
343         198.000000  3775.000000  female  2009  

[344 rows x 9 columns]